### ISSUE TITLE
[selectors] Parameterized selectors

### ISSUE THREAD

**Issue by matthewp** [w3c/csswg-drafts#10567]
CSS is great for filtering content. This lets you keep your DOM mostly static and change what is visible by toggling attributes. However, this breaks down with dynamic content.

Take this example:

```html
<input value="taco">

<ul>
  <li data-name="crunch tacos"></li>
  <li data-name="meat pie"></li>
</ul>
```

I would like to highlight the `li` that matches the input value. To achieve this today you need to write JavaScript to do the filtering. Usually this means rerendering the entire list to either remove the `li`s that don't match (which is what almost everyone does), or perhaps toggling attributes onto the matching one. In either case, this is an expensive and code-heavy way to do filtering.

If CSS selectors could be used as parameters you could achieve the filtering without manipulating the list items. Here's an example of what I would like to do: (note that i know this is horrible and invalid syntax, it's only meant to communicate the goal)

```css
input$val=attr(value) + ul [data-name^=$val] {
  color: tomato;
}
```

**Comment by noamr** [comment]
I've somewhat recently proposed something similar in the context of view transitions: https://github.com/w3c/csswg-drafts/issues/8320#issuecomment-1877522852

/cc @mirisuzanne

**Comment by matthewp** [comment]
That's much better, rewriting my thing using that syntax (if i understand it):

```css
input[:value] + ul [data-name^=[:value]] {
  color: tomato;
}
```

**Comment by ChrisShank** [comment]
I've also come across wanting something similar with expanding/collapsing groupings in HTML tables. Basically if a grouped row is collapsed then we want to hid all of the visually nested rows. (Semantically the rows in the HTML table are flat, so we want to hide the next siblings.)

```css
table tr[grouping][:path][collapsed] ~ tr[path^=[:path]] {
  display: none;
}
```

**Comment by romainmenke** [comment]
> CSS is great for filtering content. 

I would challenge this.
This pattern isn't good accessibility.

Using this pattern as the main motivation for new syntax and features seems like a bad idea.

Do you have other use cases?

Something dynamic that is hard to do today and wouldn't negatively affect accessibility if done purely with CSS?

**Comment by matthewp** [comment]
Hi @romainmenke, can you explain why this pattern is bad for accessibility? `display: none` hides the content from screen readers.

**Comment by romainmenke** [comment]
I think I was conflating two things 🤔 
I assumed that such toggling with CSS would not be good for accessibility.
But as long as you set the right aria attributes screenreaders will advertise changes.

I think I am confusing this with the old-style checkbox hacks where only the label is visible and the input itself is hidden.

Comment above edited.

**Comment by mirisuzanne** [comment]
I agree it would be useful (seemly in several situations). With @noamr's original proposal, there was a more clear distinction between the value we're grabbing (a selector syntax) and the place we're using it (in a declaration value). Once you put those both into selectors, using the same syntax, it's hard to know which is defining the parameter and which is accessing it.

I think there's been progress on disentangling the `attr()` proposal - removing the problematic bits, so that it can move forward. This feature is a bit different, but they work in a similar space, so there may be some crossover.

For this use-case specifically, the proposal only seems to address exact match. Is that good enough to solve most use-cases here?

Those are my initial reactions. I'm not an editor on the selectors spec, so it would be good to get input from @tabatkins or @fantasai as well.

**Comment by Alohci** [comment]
Selecting on an input :value pseudo-class would enable CSS keyloggers. I don't think that's a good idea.

**Comment by noamr** [comment]
> Selecting on an input :value pseudo-class would enable CSS keyloggers. I don't think that's a good idea.

It just selects the `value` attribute, not the actual typed value

**Comment by brunostasse** [comment]
One use case I can see for this feature is styling elements belonging to a component instance based on attributes set on the component instance's root element when several instances are intertwined.

The `@scope` feature, which has been partially thought to style component's elements, doesn't help in such cases, because it always ties elements to the closest component instance's root element. It doesn't allow to style elements scattered across several instances of the same component.

For example, if a button belongs to Dialog component is nested in two Dialog component instance's root element, but belongs to the outer one, then the only way to style the button based on a state/variant/modifier belonging to its component instance is to add that state/variant/modifier to the button element, and all other elements belonging to that Dialog instance. Like so:

```html
<div class="Dialog-root type-alert"> <!-- dialog-1 root element -->
   <div class="Dialog-root type-warning"> <!-- dialog-2 root element -->
      <button class="Dialog-button type-alert">Close dialog-1</button>
      <button class="Dialog-button type-warning">Close dialog-2</button>
   </div>
</div>
```
```css
.Dialog-root.type-alert {
   background-color: red;
}
.Dialog-button.type-alert {
   color: red;
}
```

Whereas with this feature, all elements belonging to one component instance can be tied together with an id, like so:

```html
<div class="Dialog-root type-alert" data-cid="dialog-1">
   <div class="Dialog-root type-warning" data-cid="dialog-2">
      <button class="Dialog-button" data-cid="dialog-1">Close dialog-1</button>
      <button class="Dialog-button" data-cid="dialog-2">Close dialog-2</button>
   </div>
</div>
```
```css
.Dialog-root[:data-cid].type-alert {
   background-color: red;

   .Dialog-button[data-cid=[:data-cid]] {
      color: red;
   }
}
```

This reduces the number of DOM operations needed to update states/variants/modifiers on a component, as it only needs to be done on its root element rather than on all affected elements.

**Comment by matthewp** [comment]
> For this use-case specifically, the proposal only seems to address exact match. Is that good enough to solve most use-cases here?

I don't think I follow, what do you mean by exact match? We have non-exact matches through the attribute selectors `^=`, `*=`, etc. I think you probably mean something different though.

----

One other use-case would be something like ranges (is that what you mean by non-exact)? For example you want to filter the selected items whose price is between 20-100. I don't know how you could express something like that though, it would mean a new type of selector I would imagine.

**Comment by DaniGuardiola** [comment]
My use case is keeping long selectors DRY, for example:

`:is(:hover,:focus-visible,[data-focus-visible]):not(disabled,[aria-disabled],[data-disabled])`

This is a selector that targets elements that are being hovered or focused through the keyboard, and are not disabled. This is common, since many times the styles will be the same for these states.

Note that it includes selectors that could be set by JavaScript libraries such as Ariakit or Radix, which typically "extend" the web platform with accessible UI primitives. These libraries can't trigger native selectors like `:focus-visible` and, therefore, need to signal them through other means like data attributes. Additionally, they might need to communicate a disabled state, but the `disabled` attribute is invalid in most elements, and therefore either `aria-disabled` or a data attribute must be set.

I would like to be able to define this once, and use it everywhere to target these states consistently. My current solution is a custom PostCSS plugin that adds states through `:is-targeted` and similar non-standard pseudo selectors, which get transformed into their long variants.

While the approach works, it results in a hard to parse output which is also larger. The alternative (writing these by hand) can result in significant usability or accessibility problems, as it's fairly easy to miss an important part of a selector or forget to update existing ones with important changes (e.g. if a library is upgraded).

I hope I explained my use-case clearly enough! Would love to see this land.

**Comment by romainmenke** [comment]
> use case is keeping long selectors DRY

Aren't custom selectors a better proposal for that?

```css
@custom-selector :--on-interaction :is(:hover,:focus-visible,[data-focus-visible]):not(disabled,[aria-disabled],[data-disabled]);

.button:--on-interaction { color: green; }
```

**Comment by DaniGuardiola** [comment]
@romainmenke yes, this exactly how I envision it (specific syntax aside). Is there already a proposal for this?

**Comment by romainmenke** [comment]
This is the relevant specification: https://drafts.csswg.org/css-extensions/#custom-selectors

**Comment by mirisuzanne** [comment]
> I don't think I follow, what do you mean by exact match? We have non-exact matches through the attribute selectors `^=`, `*=`, etc. I think you probably mean something different though.

@matthewp yeah, that provides some flexibility. I don't know that I had a specific use-case in mind, just noting the limitation. I think your range idea is a good example that would require more. I wasn't intending it as a blocker - just noting that 'search using selectors' use-case could open a wide range of more regex-like pattern-matching needs.

**Comment by bramus** [comment]
> I've somewhat recently proposed something similar in the context of view transitions: [#8320 (comment)](https://github.com/w3c/csswg-drafts/issues/8320#issuecomment-1877522852)

@noamr Your proposal captures part of the selectors to then use as values in the declarations. This proposal is different in the sense that it captures values from the selectors and uses those values in the same selector. There is a clear distinction between reading a value and actually using it – which also was a concern @mirisuzanne [raised](https://github.com/w3c/csswg-drafts/issues/10567#issuecomment-2226545200)


> For this use-case specifically, the proposal only seems to address exact match. Is that good enough to solve most use-cases here?

One use-case that popped up – and matches this assumption – is to style the element that triggers a popover.

```css
[popovertarget={id}] {
}
```

**Comment by noamr** [comment]
> > I've somewhat recently proposed something similar in the context of view transitions: [#8320 (comment)](https://github.com/w3c/csswg-drafts/issues/8320#issuecomment-1877522852)
> 
> @noamr Your proposal captures part of the selectors to then use as values in the declarations. This proposal is different in the sense that it captures values from the selectors and uses those values in the same selector. There is a clear distinction between reading a value and actually using it – which also was a concern @mirisuzanne [raised](https://github.com/w3c/csswg-drafts/issues/10567#issuecomment-2226545200)

Agreed, I saw those are somewhat related rather than identical proposal.

**Comment by matthewp** [comment]
@bramus I don't think capturing and using in the same selector is required here. With nesting you can achieve the same thing in a nested selector. For me that's satisfactory if it makes capture/use syntax easier to achieve.

**Comment by doesDesign** [comment]
if I understand this premise right, I think this would be very useful. 

for example, I have a div. Inside I have a select and it can have a value of Apple, Orange, or Pear. 

Also within that div I have sub-fields that I want to show conditionally. So if the user selects Apple two extra Apple-specific fields appear below, if they select Orange oat Orange-specific field appears, if they select Pear nothing appears.

All the fields are present in the DOM, just display:none.

I could write a has selector:

`div:has(select[value=$dynamic-value]) .subfield[parent=$dynamic-value {
display:block;
}
`

where $dynamic-value is something that matches.

it doesn’t have to be value based either, it could be data attr or even class based maybe. Like in the case of selectedcontent having a span with a specific class, show any sub fields with that class.

Today I would have to write out a selector for every possible parent value with a corresponding child. most likely I would generate this via JS, but CSS is better

**Comment by LeaVerou** [comment]
This is very powerful, but I suspect that the sheer complexity of generic backreferences will make it very hard to get implemented.

Could there be a less powerful feature that still covers a good chunk of use cases?
For idrefs, I proposed https://github.com/w3c/csswg-drafts/issues/10970 a few months ago which would address #12436 as well. What % of use cases are not idrefs? Perhaps we could extend to a syntax that takes the source attribute as a parameter as well. E.g. `/attref()/` or even just `/ref()/`. I’ll open a new issue.

**Comment by LeaVerou** [comment]
> This is very powerful, but I suspect that the sheer complexity of generic backreferences will make it very hard to get implemented.
> 
> Could there be a less powerful feature that still covers a good chunk of use cases? For idrefs, I proposed [#10970](https://github.com/w3c/csswg-drafts/issues/10970) a few months ago which would address [#12436](https://github.com/w3c/csswg-drafts/issues/12436) as well. What % of use cases are not idrefs? Perhaps we could extend to a syntax that takes the source attribute as a parameter as well. E.g. `/attref()/` or even just `/ref()/`. I’ll open a new issue.

New issue opened #12446 but in the course of writing it, I became convinced that for certain use cases, a backreference syntax like the one discussed here is going to have vastly better ergonomics. 

A combinator works well when the query is wide ("give me ALL elements whose `id` is the same as this element's `foo` attribute, wherever they are in the document") but not well when the query needs to be combined with another combinator ("give me all child elements whose `id` is the same as this element's `foo` attribute). It can be accomplished via filtering, but it's awkward.

The scope of the backreference would need to be the whole selector list for this syntax to work for the wide use cases as well though, otherwise it becomes very awkward to do the wide queries. But that breaks a fundamental assumption of selector lists, that all its constituent complex selectors are independent.

**Comment by bramus** [comment]
A simple use case came up [on social media](https://bsky.app/profile/nicr.dev/post/3lvgz4uloxk2m) again:

> When I hover over an element with a `data-foo` attribute, make all the elements that have `data-foo` _set to the same value_ red
